src/lang/builtins_source.ml

```diff
@@ -109,3 +109,35 @@ let () =
         Clock.force_init (fun x -> List.exists (fun y -> Oo.id x = Oo.id y) l)
       in
       Lang.list ~t:s_t (List.map (fun x -> Lang.source (x :> Source.source)) l))
+
+let () =
+  let log = Log.make ["source"; "dump"] in
+  let kind = Lang.univ_t () in
+  add_builtin "source.dump" ~cat:Liq
+    ~descr:"Immediately encode the whole contents of a source into a file."
+    ~flags:[Lang.Experimental]
+    [
+      ("", Lang.format_t kind, None, Some "Encoding format.");
+      ("", Lang.string_t, None, Some "Name of the file.");
+      ("", Lang.source_t (Lang.univ_t ()), None, Some "Source to encode");
+    ]
+    Lang.unit_t
+    (fun p ->
+      let proto =
+        let p = Pipe_output.file_proto (Lang.univ_t ()) in
+        List.filter_map (fun (l, _, v, _) -> Option.map (fun v -> (l, v)) v) p
+      in
+      let proto = ("fallible", Lang.bool true) :: proto in
+      let s = Lang.to_source (Lang.assoc "" 3 p) in
+      let p = (("id", Lang.string "source_dumper") :: p) @ proto in
+      let fo = Pipe_output.new_file_output p in
+      fo#get_ready [s];
+      fo#output_get_ready;
+      log#info "Start dumping source.";
+      while s#is_ready do
+        fo#output;
+        fo#after_output
+      done;
+      log#info "Source dumped.";
+      fo#leave s;
```

Comment:
Would you mind factoring this out to a `Utils`? I think this should prove pretty useful in other situations and also would make it possible to fix all cases using this trick if we later realize something needs to be changed with this flow. Thks!

Reply:
Actually, now that I'm thinking about it, how about surfacing this as a proper operator and using it instead?

`source.synchronous` -> process the whole source synchronously as in this code, and then, instead of `source.dum`, do: `source.synchronous(output.file(...))`..

Reply:
The thing is that the input source has to be an output. Can we enforce this in the type?

Reply:
Ha, good point. No, I don't think that we have that in the type. `active_source` is used to denote sources that get actively pulled but that is not specific to outputs.

I guess we should merge as is with the workflow factored out as a `Utils` and revisit later if we see a more general need for this in the future. Now that I think about it, some (most?) outputs would not make sense with `source.synchronous`, like `output.icecast` etc.